### PR TITLE
Synchronization fix

### PR DESCRIPTION
**Pull request by danpovey**
This pull request is a different fix to the problem Karel was addressing in PR #1615.
It makes the accumulation of timing stats dependent on verbose level >= 1.
@LvHang, can you please test this PR instead of #1615, since this is the one that's a candidate to merge?
However if it would be easy and you're nearly done testing PR #1615,  testing them both is fine.

Please make run the nnet training scripts in the mini-librispeech setup.  However, you don't have to run that setup from the start, just the neural net training.  When you're done, show me the location; I want to look at the logs.

### CODE REVIEW

**Comment by LvHang** [comment]
Sure. I will do that.
About PR #1615 
Today, I run the kaldi master branch and karel's branch.I test them in verbose=0(default) or verbose=2.
In karel's verbose=2 case, the "Total GPU time" is about 1s with 'Tesla M40'. Compare with karel's verbose=0 case whose "Total GPU time" is about 80s with 'Tesla K10.G2'. It's incredible.
So I prepare to reserve a gpu `qlogin -l gpu=1` and retest them in the same gpu.
Hang

**Comment by danpovey** [comment]
The printed GPU time is not actually meaningful because in that branch, the
timing stats are totally wrong.  It's the total time elapsed that you
should look at.

**Comment by LvHang** [comment]
Oh, I see.
But I still think to reserve the same gpu is necessary. Different type gpus bring a huge time elapsed gap.
For example, when I test kaldi master in verbose=0:
If I use k80, the accounting time is 63s
(in /export/a11/hlyu/kaldi/egs/mini_librispeech/s5/exp/chain/tdnn1a_sp_verbose0/log/train.4.1.log)
If I use k10, the accounting time is 92s
(in /export/a11/hlyu/kaldi/egs/mini_librispeech/s5/exp/chain/tdnn1a_sp_verbose0/log/train.2.2.log)

Assuming I reserve one gpu, which of the following two ways about testing is better you think?
(1) run all train.$iteration.$jobs and count the average time elapsed.
(2) just run train.0.1 5times, and count the average time elapsed.
Hang

**Comment by danpovey** [comment]
It's fine to run just one job, you don't have to run an entire training run.
You don't have to run it more than twice or so.

**Comment by LvHang** [comment]
Hi Dan,
I tested karel's version(in #1615 PR), kaldi master's version and your version(in this PR) separately.
For karel's version, you can get all the files in /export/a11/hlyu/cumatrix_accelerate_by_async/egs/mini_librispeech/s5/exp/chain/tdnn1a_sp_verbose{0,1,2}
For kaldi master's version, you can get all the files in
/export/a11/hlyu/kaldi/egs/mini_librispeech/s5/exp/chain/tdnn1a_sp_verbose{0,1,2}
For your version, you can get the files in
/export/a11/hlyu/synchronization_fix/egs/mini_librispeech/s5/exp/chain/tdnn1a_sp.

I checked your setup about verbose.`verbose_opt = ("--verbose=1" if iter % 20 == 0 and iter > 0 else "")`
As the mini_librispeech only has 6 iterations, so the verbose always equals 0.

Meanwhile, I run a job (train.0.1) for the three versions code in 3 verbose(0,1,2) with the same gpu(Tesla k20m) manually. (You can get the scripts and log files in /export/a11/hlyu/synchronization_test/)
dan_verbose0 time=74; dan_verbose1 time=74; dan_verbose2 time=79
karel_verbose0 time-84; karel_verbose1 time=77; karel_verbose2 time=82
master_verbose0 time=79; master_verbose1 time=79; master_verbose2 time=84

Bests,
Hang

**Comment by vesis84** [comment]
This is strange, I would expect a larger speedup. It 1-GPU training? What is it BLSTM topology?
(For feedforward nets the speedup should be smaller, there are >100x less kernels called...)
K.

**Comment by danpovey** [comment]
Try with the TDNN+LSTM script.  For TDNN we expect less difference.

**Comment by LvHang** [comment]
Yeah, the result is about TDNN script, which is similar with DNN in topology.
I will try TDNN+LSTM and BLSTM with mini_librispeech today.
Hang

**Comment by LvHang** [comment]
Hi Dan,
This is a report about nnet3/tdnn_lstm_sp with mini_librispeech.
I tested karel's version(in #1615 PR), kaldi master's version and your version(in this PR) separately.
For karel's version, you can get all the files in /export/a11/hlyu/cumatrix_accelerate_by_async/egs/mini_librispeech/s5/exp/nnet3/tdnn_lstm1b_sp_verbose{0,1,2}
For kaldi master's version, you can get all the files in
/export/a11/hlyu/kaldi/egs/mini_librispeech/s5/exp/nnet3/tdnn_lstm1b_sp_verbose{0,1,2}
For your version, you can get the files in
/export/a11/hlyu/synchronization_fix/egs/mini_librispeech/s5/exp/nnet3/tdnn_lstm1b_sp.

For nnet3/tdnn_lstm with mini_librispeech, it has 33 iterations.

Meanwhile, I run a job (train.0.1) for the three versions code in 3 verbose(0,1,2) with the same gpu(Tesla K10.G2.8GB) manually. (You can get the scripts and log files in /export/a11/hlyu/synchronization_test_tdnn_lstm/)
dan_verbose0 time=135; dan_verbose1 time=155; dan_verbose2 time=154
karel_verbose0 time=148; karel_verbose1 time=149; karel_verbose2 time=148
master_verbose0 time=158; master_verbose1 time=157; master_verbose2 time=157

Bests,
Hang

**Comment by danpovey** [comment]
Thanks.
That's a really nice speedup!
I think I'll merge this change now.  I think it's pretty harmless.

File: src/cudamatrix/cu-tp-matrix.cc [hunk 1]
```diff
@@ -64,7 +64,7 @@ void CuTpMatrix<Real>::Invert() {
 #if HAVE_CUDA==1
   if (CuDevice::Instantiate().Enabled()) {
     if (this->num_rows_ == 0) return;
-    Timer tim;
+    CuTimer tim;
```
Comment (vesis84) [inline src/cudamatrix/cu-tp-matrix.cc]: It seems that this timer is never used for logging its elapsed time...

File: src/cudamatrix/cu-tp-matrix.cc [hunk 2]
```diff
@@ -92,7 +92,7 @@ void CuTpMatrix<Real>::CopyFromMat(const CuMatrixBase<Real> &M,
     MatrixIndexT num_rows = this->num_rows_;
     KALDI_ASSERT(num_rows == M.NumRows() && this->num_rows_ == M.NumCols());
     if (num_rows == 0) return;
-    Timer tim;
+    CuTimer tim;
```
Comment (vesis84) [inline src/cudamatrix/cu-tp-matrix.cc]: It seems that this timer is never used for logging its elapsed time...

File: src/cudamatrix/cu-vector.cc [hunk 3]
```diff
@@ -617,7 +617,7 @@ void CuVectorBase<Real>::AddTpVec(const Real alpha, const CuTpMatrix<Real> &M,
 #if HAVE_CUDA == 1
   if (CuDevice::Instantiate().Enabled()) {
     if (dim_ == 0) return;
-    Timer tim;
+    CuTimer tim;
```
Comment (vesis84) [inline src/cudamatrix/cu-vector.cc]: It seems that this timer is never used for logging its elapsed time...

File: src/cudamatrix/cu-tp-matrix.cc (hunk 2 above)
Comment (danpovey) [inline src/cudamatrix/cu-tp-matrix.cc]: @LvHang, can you please make a PR to fix this?

**Comment by LvHang** [comment]
OK

**Comment by vesis84** [comment]
Hi, and here is my BLSTM speed test, the order is: 
1) Hainan's implementation,
2) My PR #1615,
3) Original code with 'device syncs',

```
1) [TRAINING, **2.16752** min, fps9571.99] exp/blstm4i_DAN_ASYNC/log/iter02.tr.log:LOG
2) [TRAINING, **2.81618** min, fps7367.23] exp/blstm4i_GEMM_SYNC_cudaGetLastError_FINAL/log/iter02.tr.log:LOG
3) [TRAINING, **5.05169** min, fps4107.03] exp/blstm4i_OLDBIN_SYNC/log/iter02.tr.log:LOG
```

**Pull request merged**